### ISSUE TITLE
Deserialized inner class does not have a reference to its enclosing outer class

### ISSUE THREAD

**Issue by gsteckman** [Kotlin/kotlinx.serialization#1101]
**Describe the bug**
When trying to serialize/deserialize a class with an inner class, upon deserialization the inner class does not have a reference to its enclosing outer class and will throw a null pointer exception when referencing a property of the outer class.

**To Reproduce**

Example class implementation:

```
@Serializable
class OuterTestClass(val x: Int){
    val inside = InnerTestClass(5)

    @Serializable
    inner class InnerTestClass(val y: Int){
        fun compute(): Int{
            return x*y
        }
    }
}
```

Example test, where a null pointer exception is thrown at the call to o2.inside.compute():

```
@Test
fun testInnerClassesAreSerializable(){
    val o1 = OuterTestClass(3)
    val str = Json.encodeToString(o1)
    val o2 = Json.decodeFromString<OuterTestClass>(str)

    assertEquals(o1.x, o2.x)
    assertEquals(o1.inside.y, o2.inside.y)
    assertEquals(15, o1.inside.compute())
    assertEquals(o1.inside.compute(), o2.inside.compute())
}
```

**Expected behavior**
The test should pass without throwing a null pointer exception. The deserialized InnerTestClass instance should be able to access the member "x" of its enclosing outer class.

**Environment**
 - Kotlin version: 1.4.10
 - Library version: Reproduced with both 1.0.0-RC and 1.0.0-RC2
 - Kotlin platforms: JVM
 - Gradle version: 6.6.1

**Comment by shanshin** [comment]
Serializable inner classes are not supported and most likely will cause compile error in future releases.
May you describe your use case for serializable inner classes?

**Comment by gsteckman** [comment]
I don't recall my specific use case at the time, and I changed the implementation to not use inner classes. At about the same time I learned serialization also doesn't support arbitrary object graphs and would cause a stack overflow at runtime if there is a cycle in the graph. If some language features, like inner classes, are intentionally unsupported it would be helpful if they were clearly documented.
